### ISSUE TITLE
restat in render in static.File raise Exception if file doesn't exists

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3634]
|<img alt="andrewmaster's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| andrewmaster reported|
|-|-|
|Trac ID|trac#3634|
|Type|defect|
|Created|2009-01-27 18:24:34Z|
|Branch|https://github.com/twisted/twisted/tree/norestat-3634|

'self.restat' called with default reraise=True in 'render' method of static.File without checking file existence and any exception handling.

<details><summary>Searchable metadata</summary>

```
trac-id__3634 3634
type__defect defect
reporter__andrewmaster andrewmaster
priority__highest highest
milestone__ 
branch__branches_norestat_3634 branches/norestat-3634
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1233080674000000 1233080674000000
changetime__1236006961000000 1236006961000000
version__None None
owner__ 
cc__andrewmaster@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [26150]) Branching to 'norestat-3634'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @glyph<br>@exarkun set status to `new`|
|-|-|

  1. The prevailing testing approach in `test_static.py` is to actually render the resource/request (using `self._render`) and then assert things about the response.  The two new tests take a different approach.  Following the existing convention would be nice.  Alternatively, filing a ticket about why the existing convention sucks and should be changed would be nice.
  1. The new tests pass if the restat calls are _deleted_ instead of told not to raise an exception.  Also, no other tests fail if the restat calls are deleted.
  1. The docstring for `test_staticFileDeletedRender` is incomplete.

Of these, you should really fix the third.  It would be nice if the first two were addressed in some way as well, but I won't require it for this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_8"></a>|@washort removed owner|
|-|-|

I've addressed [#11338](https://github.com/twisted/twisted/issues/11338) and #3.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @washort<br>@exarkun set status to `new`|
|-|-|

Only one of the two new tests is using `_render` now, so I'm not sure 1. is really addressed. :)  However, whatever.  Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_11"></a>|@washort set status to `closed`|
|-|-|

In ([26327]) Merge norestat-3634.
`twisted.web.static.File`'s `render` and `getChild` methods now call `restat` with `reraise=False`.

Author: glyph, washort
Reviewer: exarkun

Closes [#3634](https://github.com/twisted/twisted/issues/3634).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
